clang/lib/CIR/Dialect/IR/CIRTypes.cpp

```diff
@@ -96,6 +96,32 @@ void CIRDialect::printType(Type type, DialectAsmPrinter &os) const {
       });
 }
 
+Type PointerType::parse(mlir::AsmParser &parser) {
+  if (parser.parseLess())
+    return Type();
+  Type pointeeType;
+  unsigned addrSpace = 0;
+  if (parser.parseType(pointeeType))
+    return Type();
+  if (parser.parseOptionalComma().succeeded()) {
+    if (parser.parseInteger(addrSpace))
+      return Type();
+  }
+  if (parser.parseGreater())
+    return Type();
+  return get(parser.getContext(), pointeeType, addrSpace);
+}
+
+void PointerType::print(mlir::AsmPrinter &printer) const {
+  printer << '<';
+  printer.printType(getPointee());
+  if (auto addrSpace = getAddrSpace()) {
+    printer << ", ";
+    printer << addrSpace;
+  }
+  printer << '>';
+}
+
```

Comment:
I believe this syntax could be defined entirely in TableGen:

```
let assemblyFormat = [{
  `<` $pointee ( `,` $addressSpace $addrSpace^)? `>`
}];
```

Besides, I suggest to make the intention of the address space parameter clearer in the syntax:

```
let assemblyFormat = [{
  `<` $pointee ( `,` `address_space` `(` $addressSpace `)` $addrSpace^)? `>`
}];
```

Reply:
Thanks, good to know! I used to try optional token in asm format, but I doubted that I would fail to deal with the builder. Just occurred to me, LLVMPointerType can be a reference impl, in which `DefaultValuedParameter` is used. I'll try this combination first.

As for extra hint token `address_space`, it's great idea and open for discussion~ Like, we can also align it with llvm's `addrspace(i)`. Do you know any similar constructs in ClangIR for reference of naming style?

Reply:
> Do you know any similar constructs in ClangIR for reference of naming style?

AFAIK we don't have an official naming style yet. Both of `addrspace` and `addr_space` should be good.